File: gunhoflash/longest-palindromic-substring.py
class Solution:
  def isPalindrome(self, s: str) -> bool:
    return s == s[::-1]

  def longestPalindrome(self, s: str) -> str:
    s = '-' + s
    L = len(s)

    longestPalindromeStartIndex = 1
    longestPalindromeLength = 1

    for i in range(2, L):
      left = i - longestPalindromeLength
      right = i

      if self.isPalindrome(s[left - 1 : right + 1]):
        # 2 step longer palindrome found
        longestPalindromeStartIndex = left - 1
        longestPalindromeLength += 2
      elif self.isPalindrome(s[left : right + 1]):
        # 1 step longer palindrome found
        longestPalindromeStartIndex = left
        longestPalindromeLength += 1
    
    return s[longestPalindromeStartIndex : longestPalindromeStartIndex + longestPalindromeLength]

File: gunhoflash/longest-palindromic-substring.ts
function isPalindrome(s: string): boolean {
  const L = s.length;
  const half = (L + 1) >> 1;
  for (let i = 0; i < half; i++) {
    if (s[i] !== s[length - 1 - i]) {
      return false;
    }
  }
  return true;
}

function longestPalindrome(s: string): string {
  s = '-' + s;
  const L = s.length;

  let longestPalindromeStartIndex = 1;
  let longestPalindromeLength = 1;

  for (let i = 2; i < L; i++) {
    const left = i - longestPalindromeLength;
    const right = i;

    if (isPalindrome(s.substring(left - 1, right + 1))) {
      // 2 step longer palindrome found
      longestPalindromeStartIndex = left - 1;
      longestPalindromeLength += 2;
    } else if (isPalindrome(s.substring(left, right + 1))) {
      // 1 step longer palindrome found
      longestPalindromeStartIndex = left;
      longestPalindromeLength += 1;
    }
  }

  return s.substring(
    longestPalindromeStartIndex,
    longestPalindromeStartIndex + longestPalindromeLength
  );
}
